src/LinuxLudo.API/Hubs/GameHub.cs

```diff
@@ -25,19 +24,74 @@ public async Task JoinGame(string username, Guid gameId)
 
+            // Add the new player to the joined game's group
+            await Groups.AddToGroupAsync(Context.ConnectionId, gameId.ToString());
+
             // Updates all clients with the latest player list
-            await SendJoinGame(username, _repository.FetchGameById(gameId).PlayersInGame);
+            await SendConnectionChanged(gameId.ToString(), username, _repository.FetchGameById(gameId).PlayersInGame);
+
+            // Update the specific player on whose turn it is
+            await Clients.Client(Context.ConnectionId).SendAsync("ReceivePlayerTurn", game.CurrentTurnColor);
+        }
+
+        public async Task RollDice(string username)
+        {
+            // Fetches the game the player is in
+            OpenGame game = _repository.FetchGameById(_repository.FetchUserById(Context.ConnectionId).JoinedGame.GameId);
+            int roll = engine.RollDice();
+
+            // Notify clients of roll
+            await Clients.Group(game.GameId.ToString()).SendAsync("ReceiveRollDice", username, roll);
+
+            // Set the turn to the next player
+            await UpdatePlayerTurn(game);
+        }
+
+        private async Task UpdatePlayerTurn(OpenGame game)
+        {
+            // Sets the current turn to the next player in order
+            game.CurrentTurnColor = engine.UpdatePlayerTurn(game);
+
+            // Updates the clients with whose turn it is
+            await Clients.Group(game.GameId.ToString()).SendAsync("ReceivePlayerTurn", game.CurrentTurnColor);
+        }
+
+        private async Task SendConnectionChanged(string gameId, string player, List<Player> players)
+        {
+            // Broadcasts to all connected clients (users) that a new player has joined
+            await Clients.Group(gameId).SendAsync("ReceiveConnectionChanged", player, players);
         }
 
-        private async Task SendJoinGame(string player, List<Player> players)
+        public override async Task OnDisconnectedAsync(Exception exception)
         {
-            await Clients.All.SendAsync("ReceiveJoinGame", player, players);
+            ConnectedUser user = _repository.FetchUserById(Context.ConnectionId);
+            if (user?.JoinedGame != null)
+            {
+                // If it's the user that disconnected's turn, change turn to the next one
+                if (user.JoinedGame.CurrentTurnColor == user.JoinedGame.PlayersInGame.First(player => player.Name == user.Username).Color)
+                {
+                    await UpdatePlayerTurn(user.JoinedGame);
+                }
+
+                // Removes the player from their game
+                _repository.RemovePlayer(user.JoinedGame, user.Username);
+
+                // Update the clients that user has left
+                await SendConnectionChanged(user.JoinedGame.GameId.ToString(), user.Username, _repository.FetchGameById(user.JoinedGame.GameId).PlayersInGame);
+
+            }
+
+            await base.OnDisconnectedAsync(exception);
```

Comment:
To make it abit easier to read and understand try to switch the if stmt around
```suggestion
            if (user?.JoinedGame == null)
            {
                await base.OnDisconnectedAsync(exception);
                return
            }

            if (user.JoinedGame.CurrentTurnColor == user.JoinedGame.PlayersInGame.First(player => player.Name == user.Username).Color)
            {
               await UpdatePlayerTurn(user.JoinedGame);
            }

            _repository.RemovePlayer(user.JoinedGame, user.Username);

            await SendConnectionChanged(user.JoinedGame.GameId.ToString(), user.Username, _repository.FetchGameById(user.JoinedGame.GameId).PlayersInGame);
            await base.OnDisconnectedAsync(exception);
```

Reply:
Repository RemovePlayer method cannot be called if user.Joined game is null. - Leaving as is for now.